pandas/core/categorical.py

```diff
@@ -435,10 +436,24 @@ def astype(self, dtype, copy=True):
             .. versionadded:: 0.19.0
 
         """
+        if isinstance(dtype, compat.string_types) and dtype == 'category':
+            # GH 18593: astype('category') should not change anything
+            return self.copy() if copy else self
+
+        dtype = pandas_dtype(dtype)
         if is_categorical_dtype(dtype):
-            if copy is True:
-                return self.copy()
-            return self
+            # GH 18593: keep current categories if None (ordered can't be None)
+            if dtype.categories is None:
+                new_categories = self.categories
```

Comment:
if you also set ``ordered=False`` here (for ``dtype.categories`` is None) and the else take the ordered from the dtype, then I believe you can remvoe 439-441 (also needt o make 450 be

``dtype = CategoricalDtype(new_categories, ordered)``

Reply:
The problem is that `astype('category')` should not change the ordered attribute (so not set it always to False), so you would need to take the `ordered` from `self`. But then, you are not really sure if the user specified the order of the `CategoricalDtype` specifically, or if the `ordered=False` came from the default value. 

To summarize, I think it is easier to leave it as is and treat `'category'` as a special case.

(we might want to check if we can't let the ordered keyword have a default of None to make it easier to deal with this, but that is another issue)

Reply:
so this is pretty much the same block of code as in #18677 

this is way too special casey to exist in 2 places. Can you refator this to a common method that you can use in #18677 (or on that PR is ok too).

I'd like to do that before merging either one of these.

Reply:
> The problem is that astype('category') should not change the ordered attribute (so not set it always to False), so you would need to take the ordered from self. But then, you are not really sure if the user specified the order of the CategoricalDtype specifically, or if the ordered=False came from the default value.

not sure this is True. ``.astype('category')`` is clearly == ``CategoricalDtype()`` which by-definition has ``ordered=False``. I don't know how you can have any other conclusion. Furthermore if this is NOT the case. Then we should immediately fix this. As a special case for this is monumentally confusing. The very fact that we have to have this discussion attests to this.

Reply:
Because that would be changing the existing behaviour:

```
In [2]: pd.Categorical(['a', 'b'], ordered=True)
Out[2]: 
[a, b]
Categories (2, object): [a < b]

In [3]: pd.Categorical(['a', 'b'], ordered=True).astype('category')
Out[3]: 
[a, b]
Categories (2, object): [a < b]
```

I personally think the above is the logical behaviour, but I can also see a point in to make the above ordered=False. 
Main reason for liking the above is that 'category' == `CategoricalDtype()` and that `CategoricalDtype` has a default of `ordered=False` should be more an implementation detail to the user.

But let's maybe open a new issue to discuss that? 
And keep this PR just fixing the bug without changing existing behaviour.

Reply:
>Can you refactor this to a common method that you can use in #18677 (or on that PR is ok too)

Will do that over in #18677, since it seems to be closer to being complete.  Or can close the two individual PR's and create a new PR that combines both, if that would be preferable.  Didn't realize the fixes would be so similar until the first PR was already in review.